File: app/src/main/java/com/ne1c/gitteroid/ui/fragments/BaseFragment.java
package com.ne1c.gitteroid.ui.fragments;

import android.app.Activity;
import android.os.Bundle;
import android.support.v4.app.Fragment;

import com.ne1c.gitteroid.Application;
import com.ne1c.gitteroid.di.HasComponent;
import com.ne1c.gitteroid.di.components.ApplicationComponent;

public abstract class BaseFragment extends Fragment {
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        initDiComponent();
    }

    protected abstract void initDiComponent();

    public <T> T getActivityComponent(Class<T> clazz) {
        Activity activity = getActivity();
        HasComponent<T> has = (HasComponent<T>) activity;
        return has.getComponent();
    }

    public ApplicationComponent getAppComponent() {
        return ((Application) getActivity().getApplication()).getComponent();
    }
}

File: app/src/main/kotlin/com/ne1c/gitteroid/ui/fragments/BaseFragment.kt
package com.ne1c.gitteroid.ui.fragments

import android.app.Activity
import android.os.Bundle
import android.support.v4.app.Fragment

import com.ne1c.gitteroid.GitteroidApplication
import com.ne1c.gitteroid.di.HasComponent
import com.ne1c.gitteroid.di.components.ApplicationComponent

abstract class BaseFragment : Fragment() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initDiComponent()
    }

    protected abstract fun initDiComponent()

    fun <T> getActivityComponent(clazz: Class<T>): T {
        val activity = activity
        val has = activity as HasComponent<T>
        return has.component
    }

    val appComponent: ApplicationComponent
        get() = (activity.application as GitteroidApplication).component
}
